### ISSUE TITLE
datalad remove --recursive is not recursive really

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#1573]
seems to work to some level (e.g. if I install only ///labs/haxby) but not entirely:
```shell

$> datalad -l warning install -r ///labs  &&   datalad -l warning remove -r labs 
install(ok): /tmp/labs (dataset)               
install(ok): gobbini (dataset) [Installed subdataset <Dataset path=/tmp/labs/gobbini>]
install(ok): gobbini/famface (dataset) [Installed subdataset <Dataset path=/tmp/labs/gobbini/famface>]
install(ok): gobbini/famface/data (dataset) [Installed subdataset <Dataset path=/tmp/labs/gobbini/famface/data>]
install(ok): haxby (dataset) [Installed subdataset <Dataset path=/tmp/labs/haxby>]
install(ok): haxby/raiders (dataset) [Installed subdataset <Dataset path=/tmp/labs/haxby/raiders>]
install(ok): mitchell (dataset) [Installed subdataset <Dataset path=/tmp/labs/mitchell>]
install(ok): mitchell/science2008 (dataset) [Installed subdataset <Dataset path=/tmp/labs/mitchell/science2008>]
install(ok): tarr (dataset) [Installed subdataset <Dataset path=/tmp/labs/tarr>]
install(ok): tarr/face_place (dataset) [Installed subdataset <Dataset path=/tmp/labs/tarr/face_place>]
action summary:
  install (ok: 10)
datalad -l warning install -r ///labs  6.73s user 1.97s system 35% cpu 24.517 total
remove(notneeded): /tmp/labs/tarr [no annex'ed content]
[ERROR  ] to be uninstalled dataset <Dataset path=/tmp/labs/tarr> has present subdatasets, forgot --recursive? [remove(/tmp/labs/tarr)] 
remove(error): /tmp/labs/tarr (dataset) [to be uninstalled dataset <Dataset path=/tmp/labs/tarr> has present subdatasets, forgot --recursive?]
remove(notneeded): /tmp/labs/mitchell [no annex'ed content]
[ERROR  ] to be uninstalled dataset <Dataset path=/tmp/labs/mitchell> has present subdatasets, forgot --recursive? [remove(/tmp/labs/mitchell)] 
remove(error): /tmp/labs/mitchell (dataset) [to be uninstalled dataset <Dataset path=/tmp/labs/mitchell> has present subdatasets, forgot --recursive?]
remove(notneeded): /tmp/labs/haxby [no annex'ed content]
[ERROR  ] to be uninstalled dataset <Dataset path=/tmp/labs/haxby> has present subdatasets, forgot --recursive? [remove(/tmp/labs/haxby)] 
remove(error): /tmp/labs/haxby (dataset) [to be uninstalled dataset <Dataset path=/tmp/labs/haxby> has present subdatasets, forgot --recursive?]
remove(notneeded): /tmp/labs/gobbini [no annex'ed content]
[ERROR  ] to be uninstalled dataset <Dataset path=/tmp/labs/gobbini> has present subdatasets, forgot --recursive? [remove(/tmp/labs/gobbini)] 
remove(error): /tmp/labs/gobbini (dataset) [to be uninstalled dataset <Dataset path=/tmp/labs/gobbini> has present subdatasets, forgot --recursive?]
remove(notneeded): /tmp/labs [no annex'ed content]
[ERROR  ] to be uninstalled dataset <Dataset path=/tmp/labs> has present subdatasets, forgot --recursive? [remove(/tmp/labs)] 
remove(error): /tmp/labs (dataset) [to be uninstalled dataset <Dataset path=/tmp/labs> has present subdatasets, forgot --recursive?]
action summary:
  remove (error: 5, notneeded: 5)

```

**Comment by bpoldrack** [comment]
Hm. The issue seems to not be the recursion, but that those subdatasets are marked as `notneeded`.

**Comment by yarikoptic** [comment]
indeed, then I guess logic is flawed and `, forgot --recursive?` is just misleading and shouldn't be printed if recursive is True ;)

**Comment by mih** [comment]
That message is old, I just kept it. The logic isn't completely flawed either, but it confuses the states of actions. The 'notneeded' is coming from the internal `drop` call (plain git, drop not needed), but it should continue removing afterwards, which it doesn't. I'll look into it.

**Comment by mih** [comment]
Yeah, so it is really fucked up -- will take a while to understand what is happening.

**Comment by yarikoptic** [comment]
hm... that is quite sad since then I couldn't demo anything if I would need to "sudo rm -rf " all the time instead of using a proper datalad command.

**Comment by yarikoptic** [comment]
I have sincerely tried to comprehend all the switches given to AnnotatePaths but was not sure even which knob I could potentially tune there... so I really hope that you might figure out the mystery on a fresh start ;)

**Comment by yarikoptic** [comment]
FWIW -- this was the "minimalistic" testcase I was testing on
```python
@skip_if_no_network
@with_tempfile(mkdir=True)
@use_cassette('test_remove_recursive_2')
def test_remove_recursive_2(tdir):
    # fails in some cases https://github.com/datalad/datalad/issues/1573
    with chpwd(tdir):
        install('///labs')
        install('labs/tarr/face_place')
        remove('labs', recursive=True)
```

**Comment by yarikoptic** [comment]
really not sure why we have `ds.uninstall` and `_uninstall_dataset` and which I should use when... then I thought that 
```
@@ -182,8 +183,21 @@ class Remove(Interface):
                                 uninstall_failed = True
                             yield r
                     if ap['path'] == ds_path:
+                        if recursive:
+                            # First uninstall all the present subdatasets
+                            for subds_ in ds.subdatasets(fulfilled=True):
+                                for r in ds.uninstall(
+                                        subds_['path'], recursive=recursive, check=check,
+                                        if_dirty=if_dirty, result_xfm=None,
+                                        result_filter=None,
+                                        on_failure='ignore'
+                                ):
+                                    pass  # yield r
```
would do it but it also didn't ...

**Comment by mih** [comment]
Fixes coming in #1575 still not at the bottom of this, but found several bugs already.

**Comment by mih** [comment]
I am about to push the changes (massive set). But the result is this:

````
mih@meiner /tmp/datasets.datalad.org (git)-[master] % datalad -l warning remove -r labs
drop(notneeded): /tmp/datasets.datalad.org/labs/tarr/face_place (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/tarr/face_place]
uninstall(ok): /tmp/datasets.datalad.org/labs/tarr/face_place (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/tarr [no annex'ed content]
uninstall(ok): /tmp/datasets.datalad.org/labs/tarr (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/mitchell/science2008 (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/mitchell/science2008]
uninstall(ok): /tmp/datasets.datalad.org/labs/mitchell/science2008 (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/mitchell [no annex'ed content]
uninstall(ok): /tmp/datasets.datalad.org/labs/mitchell (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/haxby/raiders (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/haxby/raiders]
uninstall(ok): /tmp/datasets.datalad.org/labs/haxby/raiders (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/haxby [no annex'ed content]
uninstall(ok): /tmp/datasets.datalad.org/labs/haxby (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/gobbini/famface/data (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/gobbini/famface/data]
uninstall(ok): /tmp/datasets.datalad.org/labs/gobbini/famface/data (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/gobbini/famface [no annex'ed content]
uninstall(ok): /tmp/datasets.datalad.org/labs/gobbini/famface (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs/gobbini [no annex'ed content]
uninstall(ok): /tmp/datasets.datalad.org/labs/gobbini (dataset)
drop(notneeded): /tmp/datasets.datalad.org/labs [no annex'ed content]
remove(ok): /tmp/datasets.datalad.org/labs (dataset)
action summary:
  drop (notneeded: 10)
  remove (ok: 1)
  uninstall (ok: 9)
````

Much cleaner than it was before. Importantly, this is now functionally equivalent to:

````
mih@meiner /tmp/datasets.datalad.org (git)-[master] % datalad -l warning remove -d labs -r
drop(notneeded): tarr/face_place (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/tarr/face_place]
uninstall(ok): tarr/face_place (dataset)
drop(notneeded): tarr [no annex'ed content]
uninstall(ok): tarr (dataset)
drop(notneeded): mitchell/science2008 (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/mitchell/science2008]
uninstall(ok): mitchell/science2008 (dataset)
drop(notneeded): mitchell [no annex'ed content]
uninstall(ok): mitchell (dataset)
drop(notneeded): haxby/raiders (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/haxby/raiders]
uninstall(ok): haxby/raiders (dataset)
drop(notneeded): haxby [no annex'ed content]
uninstall(ok): haxby (dataset)
drop(notneeded): gobbini/famface/data (directory) [nothing to drop from /tmp/datasets.datalad.org/labs/gobbini/famface/data]
uninstall(ok): gobbini/famface/data (dataset)
drop(notneeded): gobbini/famface [no annex'ed content]
uninstall(ok): gobbini/famface (dataset)
drop(notneeded): gobbini [no annex'ed content]
uninstall(ok): gobbini (dataset)
drop(notneeded): . [no annex'ed content]
remove(ok): . (dataset)
action summary:
  drop (notneeded: 10)
  remove (ok: 1)
  uninstall (ok: 9)
````
